src/Quartz.Spi.MongoDbJobStore/LockManager.cs

```diff
@@ -24,6 +24,8 @@ internal class LockManager : IDisposable
         private readonly ConcurrentDictionary<LockType, LockInstance> _pendingLocks =
             new ConcurrentDictionary<LockType, LockInstance>();
 
+        private readonly SemaphoreSlim _pendingLocksSemaphore = new SemaphoreSlim(1);
```

Comment:
The locking system currently in place is there for a reason. It has to be a distributed lock so that other running instances in different processes can't write or take a job that another process has already taken. This is a requirement of Quartz and since Mongo doesn't have a native record locking mechanism, this methodology relies on the atomic nature of writes in Mongo to achieve something similar. You are welcome to improve the distributed lock, but it has to remain distributed.

Reply:
When all this was written, Mongo did not have transactions so those may be useful in this regard, but that would be a major overhaul.